languages/python/sqlalchemy-oso/sqlalchemy_oso/async_session.py

```diff
@@ -0,0 +1,144 @@
+"""SQLAlchemy async session classes and factories for oso."""
+from typing import Any, Callable, Optional, Type
+import logging
+
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy_oso.session import Permissions, AuthorizedSessionBase
+from sqlalchemy import orm
+
+from oso import Oso
+
+logger = logging.getLogger(__name__)
+
+
+def async_authorized_sessionmaker(
+    get_oso: Callable[[], Oso],
+    get_user: Callable[[], Any],
+    get_checked_permissions: Callable[[], Permissions],
+    class_: Type[AsyncSession] = None,
+    **kwargs,
+):
+    """AsyncSession factory for sessions with Oso authorization applied.
+
+    :param get_oso: Callable that returns the Oso instance to use for
+                    authorization.
+    :param get_user: Callable that returns the user for an authorization
+                     request.
+    :param get_checked_permissions: Callable that returns an optional map of
+                                    permissions (resource-action pairs) to
+                                    authorize for the session. If the callable
+                                    returns ``None``, no authorization will
+                                    be applied to the session. If a map of
+                                    permissions is provided, querying for
+                                    a SQLAlchemy model present in the map
+                                    will authorize results according to the
+                                    action specified as the value in the
+                                    map. E.g., providing a map of ``{Post:
+                                    "read", User: "view"}`` where ``Post`` and
+                                    ``User`` are SQLAlchemy models will apply
+                                    authorization to ``session.query(Post)``
+                                    and ``session.query(User)`` such that
+                                    only ``Post`` objects that the user can
+                                    ``"read"`` and ``User`` objects that the
+                                    user can ``"view"`` are fetched from the
+                                    database.
+    :param class_: Base class to use for sessions.
+
+    All other keyword arguments are passed through to
+    :py:func:`sqlalchemy.orm.session.sessionmaker` unchanged.
+
+    **Invariant**: the values returned by the `get_oso()`, `get_user()`, and
+    `get_checked_permissions()` callables provided to this function *must
+    remain fixed for a given session*. This prevents authorization responses
+    from changing, ensuring that the session's identity map never contains
+    unauthorized objects.
+
+    NOTE: _baked_queries are disabled on SQLAlchemy 1.3 since the caching
```

Comment:
We can drop this note since this API will only be used on 1.4.

Reply:
Removed